### PR TITLE
Game logic feature

### PR DESCRIPTION
**Pull request by SuyashFowdar**
## Tasks done within feature

- Divide all logics in classes
- Add methods for each class
- Put each class in separate file
- Implement the winning logic

Thank You.

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    puts '   1 2 3 '
+    puts '  -------'
+    puts "a |#{@play[0][0]}|#{@play[0][1]}|#{@play[0][2]}|"
+    puts '  -------'
+    puts "b |#{@play[1][0]}|#{@play[1][1]}|#{@play[1][2]}|"
+    puts '  -------'
+    puts "c |#{@play[2][0]}|#{@play[2][1]}|#{@play[2][2]}|"
+    puts '  -------'
+    puts
+  end
```
Comment (jubaan) [inline lib/game.rb]: - [ ] There should be no puts/print outside `bin/main`

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  def check_valid_input(input, player)
+    if input.length == 2
+      @current_player = player
+      update_d(input)
+    else
+      puts 'Wrong value entered! Please try again!'
+      puts
```
Comment (jubaan) [inline lib/game.rb]: - [ ] There should be no puts/print outside `bin/main`

File: lib/game.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    check_valid_move
+  end
+
+  def check_valid_move
+    if @play[@first_d][@second_d] != 'x' && @play[@first_d][@second_d] != 'o'
+      play_move
+    else
+      puts 'This square is already filled! Please try again!'
+      puts
```
Comment (jubaan) [inline lib/game.rb]: - [ ] There should be no puts/print outside `bin/main`

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    game_won if winning_possibilities.any? do |item|
+      item.push(symbol)
+      item.uniq.length == 1
+    end
+  end
+
+  def game_won
+    game_over
+    print_game
+    puts "#{@current_player.fetch_name} has won the game!"
```
Comment (jubaan) [inline lib/game.rb]: - [ ] There should be no puts/print outside `bin/main`

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def fetch_name
+    @name
+  end
+
+  def fetch_symbol
+    @symbol
+  end
+end
```
Comment (jubaan) [inline lib/player.rb]: - [ ] Use accessor methods for `@name` and `@symbol`

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  end
+
+  def game_over
+    @gameover = true
+    puts 'Game Over!'
```
Comment (jubaan) [inline lib/game.rb]: - [ ] There should be no puts/print outside `bin/main`

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    @move_num
+  end
+
+  def update_first_d=(new_d)
+    @first_d = new_d
+  end
+
+  def update_second_d=(new_d)
+    @second_d = new_d
+  end
```
Comment (jubaan) [inline lib/game.rb]: - [ ] Use descriptive and meaningful names for methods.

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+  def update_d(input)
+    moves_first_d = %w[a b c]
+    moves_second_d = %w[1 2 3]
+
+    moves_first_d.each_with_index { |item, i| @first_d = i if input.include?(item) }
+    moves_second_d.each_with_index { |item, i| @second_d = i if input.include?(item) }
+
+    check_valid_move
+  end
```
Comment (jubaan) [inline lib/game.rb]: - [ ] Use descriptive and meaningful names for methods.

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    @move_num = 1
+  end
+
+  def fetch_gameover
+    @gameover
+  end
+
+  def fetch_move_num
+    @move_num
+  end
```
Comment (jubaan) [inline lib/game.rb]: - [ ] Use accessor methods.

File: lib/game.rb (hunk 7 above)
Comment (jubaan) [inline lib/game.rb]: - [ ] Use accessor methods.

**Review by jubaan — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## :wave: :wave: Hello @krishnzzz  and @ArreyTabe  . Thanks for submitting you project!

Great thing this project has and has done well:
:100: The game as for user names
:100: The game ends when it finds a winner
:100: The game ends when is a draw
:100: Switched between player turns

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [x] The game exits and throws an error when invalid input is entered:
`check_valid_move': no implicit conversion from nil to integer (TypeError)`

- [x] The game doesn't throw a message notifying the game is a draw.

- [x] There should be no puts/print outside `bin/main`. See inline comments.

- [x] Use accessor methods. See inline comments.


### [OPTIONAL]  :yellow_circle:  

- [ ] The game should check the player's name input. Right now the game can be played without none of the player's names, making it difficult to follow the turns of each round of the game.

## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

**Review by umairarshadbutt — APPROVED** [review]
# **Status: PROJECT APPROVED** :champagne:
Excellent job on this project guys. :fireworks: Keep it up.
##
**All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete.**
- Congratulations on the splendid job @krishnzzz  and @ArreyTabe ! :+1: 
- You can now merge to master and submit the project completion form :rocket: . I wish you all the very best in your next tasks :1st_place_medal:
##
**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Pull request merged**